doc/designs/quic-design/server/quic-polling.md

```diff
@@ -0,0 +1,1637 @@
+   signal *readiness* rather than *completion*. In fact, this is what the
+   `select` and `WSAPoll` functions use internally. Unlike those functions, this
+   is based around registering sockets in advice and submits readiness
+   notifications to an IOCP, so this can be quite performant.
+
+   `IOCTL_AFD_POLL` is an internal, undocumented API. It is however widely used,
+   and is now the basis of libuv (the I/O library used by Node.js), ZeroMQ, and
+   Rust's entire asynchronous I/O ecosystem on Windows. In other words, while
+   officially being undocumented and internal, it has in practice become widely
+   used by third-party software, to the point where it cannot really be changed
+   in future without breaking massive amounts of software. `IOCTL_AFD_POLL` has
+   been around since at least NT 4 and is supported by Wine. Moreover it is
+   worth noting that the reason why so many projects have resorted to using this
+   API on Windows is due to the sheer lack of anything providing the appropriate
+   functionality in the public API. The high level of reliance on this
+   functionality in contemporary software doing asynchronous I/O does give
+   reasonable confidence in using this API.
+
+An immediate mode interface can be implemented using option 1.
+
+Based on the above, options 1, 2 and 5 are viable for implementation of a
+retained mode interface, with option 2 being a fairly substantial hack and
+option 5 being the preferred approach for projects wanting an epoll/kqueue-style
+model on Windows. The suggested approach is therefore to implement option 5,
+though option 1 is also a viable fallback.
+
+In any case, it appears the poller API as designed and proposed above
+can be implemented adequately on Windows.
+
+Extra features on QUIC objects
+------------------------------
+
+### Low-watermark functionality
+
+Sometimes an application knows it does not need to do anything until at least N
+bytes are available to read or write. In conventional Berkeley sockets APIs this
+is known as “low-watermark” (LOWAT) functionality.
+
+Rather than making polling interfaces more convoluted by adding fields to
+polling-related structures, we propose to add a knob which can be configured on
+an individual QUIC stream:
+
+```c
+#define SSL_LOWAT_FLAG_ONESHOT     (1U << 0)
+
+int SSL_set_read_lowat(SSL *ssl, size_t loawt, uint64_t flags);
+int SSL_get_read_lowat(SSL *ssl, size_t *lowat);
+
+int SSL_set_write_lowat(SSL *ssl, size_t lowat, uint64_t flags);
+int SSL_get_write_lowat(SSL *ssl, size_t *lowat);
+```
+
+If `ONESHOT` is set, the low-watermark condition is automatically cleared
+after the next call to a read or write function respectively. The low-watermark
+condition can also be cleared by passing a low-watermark of 0.
+
+If low-watermark mode is configured, a poller will not report a stream as having
+data ready to read, or room to write data, if the amount of room available is
+less than the configured watermark.
```

Comment:
I'm not really sure we need this. At least not initially.

Reply:
Yes &mdash; just exploring possible future directions and how they would work.